src/condor_utils/file_transfer.cpp

```diff
@@ -189,7 +189,7 @@ class FileTransferItem {
 		}
 		if (is_dest_url) {
 			if (m_dest_scheme == other.m_dest_scheme) {
-				return m_dest_url < other.m_dest_url;
```

Comment:
While I can't think of anything that would break (e.g., in `std::sort`) from having a partial ordering instead of a total ordering, it's not obvious at all why this is an improvement.  You mention in the commit message it "prevents `DoUpload()` from screwing up" but that's rather vague.  Can you please add a source code comment here explaining why we shouldn't sort files by name?

Reply:
I do not presently have a pithy explanation; if I had, I would've written it up rather than asking for a conversation with the reviewer.  If the conversation generates a pithy explanation, I'll be sure to include it in the code; if the explanation remains non-pithy, I will link to it from the code instead.

Reply:
I don't think we're looking for a short explanation here (there are multiple several-hundred-word comments in the code base).

In fact, it seems quite strange on its face that an algorithm is sensitive to the sorting function when the user controls the inputs.  If having a total sorting routine breaks sometimes, what happens when the user gives the exact same ordering to the partial sort?

Reply:
Actually, re-reading the JIRA ticket, the explanation is all there -- some of the transfer items are effectively a  `mkdir` which *must* be ordered prior to the subsequent file transfer.  So, this is an attempt to preserve the ordering provided when the directory is expanded.  Not saying that you need to copy/paste the entire comment into the code ... but feel free to copy/paste liberally!

However, that said:
1. Not clear why this wasn't true in the first place.  I'm assuming it's something to do with multiple `/` characters in the generated directory filename that's not in the transfer file?
2. Later on, we use `std::sort` which can potentially reorder any equivalent items (here, we've made all non-URLs equivalent).  So, we are relying on implementation-defined behavior if this works.  I think we want to use `std::stable_sort`?

Reply:
1. If a directory only has one "source", this problem can't occur because the directory is shorter than any files in it, and has the same prefix, so it will always sort first.  In practice, the easiest way for a directory to have more than one "source" is to have the same directory in the input file transfer file list and in the checkpoint/output file transfer list... and until recently, a different bug (HTCONDOR-583) meant HTCondor would ignore the directory in SPOOL.
2. I'd already added this to the PR.

Reply:
Item 1: Ok, let's get some of this knowledge inline and I think we're ready to go.

Item 2: Sorry, missed that; I think I looked at an earlier commit and not the branch as a whole.

Reply:
I just pushed a commit with a comment that's longer than I'm happy about but I think does explain what's going on with respect to this function.  Let me know what your think.

Reply:
Ah, comment is fantastic!  Looks good to me.